src/detect-util-uint.c

```diff
@@ -0,0 +1,239 @@
+            }
+            arg3 = (char *) str_ptr;
+            SCLogDebug("Arg3 \"%s\"", arg3);
+        }
+    }
+
+    u32d = SCMalloc(sizeof (DetectU32Data));
+    if (unlikely(u32d == NULL))
+        goto error;
+
+    if (arg2 != NULL) {
+        /*set the values*/
+        switch(arg2[0]) {
+            case '<':
+            case '>':
+                if (arg3 == NULL)
+                    goto error;
+
+                if (ByteExtractStringUint32(&u32d->arg1, 10, strlen(arg3), arg3) < 0) {
+                    goto error;
+                }
+
+                SCLogDebug("u32 is %"PRIu32"",u32d->arg1);
+                if (strlen(arg1) > 0)
+                    goto error;
+
+                if (arg2[0] == '<') {
+                    u32d->mode = DETECT_UINT_LT;
+                } else { // arg2[0] == '>'
+                    u32d->mode = DETECT_UINT_GT;
+                }
+                break;
+            case '-':
+                if (arg1 == NULL || strlen(arg1)== 0)
+                    goto error;
+                if (arg3 == NULL || strlen(arg3)== 0)
+                    goto error;
+
+                u32d->mode = DETECT_UINT_RA;
+                if (ByteExtractStringUint32(&u32d->arg1, 10, strlen(arg1), arg1) < 0) {
+                    goto error;
+                }
+                if (ByteExtractStringUint32(&u32d->arg2, 10, strlen(arg3), arg3) < 0) {
+                    goto error;
+                }
+
+                SCLogDebug("u32 is %"PRIu32" to %"PRIu32"",u32d->arg1, u32d->arg2);
+                if (u32d->arg1 >= u32d->arg2) {
+                    SCLogError(SC_ERR_INVALID_SIGNATURE, "Invalid u32 range. ");
+                    goto error;
+                }
+                break;
+            default:
+                u32d->mode = DETECT_UINT_EQ;
+
+                if ((arg2 != NULL && strlen(arg2) > 0) ||
+                    (arg3 != NULL && strlen(arg3) > 0) ||
+                    (arg1 == NULL ||strlen(arg1) == 0))
+                    goto error;
+
+                if (ByteExtractStringUint32(&u32d->arg1, 10, strlen(arg1), arg1) < 0) {
+                    goto error;
+                }
+        }
+    } else {
+        u32d->mode = DETECT_UINT_EQ;
+
+        if ((arg3 != NULL && strlen(arg3) > 0) ||
+            (arg1 == NULL ||strlen(arg1) == 0))
+            goto error;
+
+        if (ByteExtractStringUint32(&u32d->arg1, 10, strlen(arg1), arg1) < 0) {
+            goto error;
+        }
+    }
+
+    SCFree(arg1);
+    SCFree(arg2);
+    SCFree(arg3);
+    return u32d;
+
+error:
+    if (u32d)
+        SCFree(u32d);
+    if (arg1)
+        SCFree(arg1);
+    if (arg2)
+        SCFree(arg2);
+    if (arg3)
+        SCFree(arg3);
+    return NULL;
+}
+
+void
+PrefilterPacketU32Set(PrefilterPacketHeaderValue *v, void *smctx)
+{
+    const DetectU32Data *a = smctx;
+    v->u8[0] = a->mode;
+    v->u32[1] = a->arg1;
+    // not enough place for a->arg2 in PrefilterPacketHeaderValue
```

Comment:
Is there anything better to do ?

Reply:
2 options I think:

1. mtu is actually u16? So then there is enough space.
2. expand space. It is currently included in:
```c
struct PrefilterPacketHeaderCtx_ {
	PrefilterPacketHeaderValue v1;                   /*     0     8 */
	uint16_t                   type;                 /*     8     2 */
	uint16_t                   value;                /*    10     2 */
	uint32_t                   sigs_cnt;             /*    12     4 */
	uint32_t *                 sigs_array;           /*    16     8 */

	/* size: 24, cachelines: 1, members: 5 */
	/* last cacheline: 24 bytes */
};
```
If we double the space this PrefilterPacketHeaderCtx_ becomes a nice half cache line

Reply:
> mtu is actually u16? So then there is enough space.

ICMPv6 MTU is u32 : https://github.com/OISF/suricata/blob/master/src/decode-icmpv6.h#L153

Reply:
So, you want to add `PrefilterPacketHeaderValue v2;` ?

Reply:
The structure passed as an argument to `PrefilterPacketU32Set` is `PrefilterPacketHeaderValue` and not its container `PrefilterPacketHeaderCtx_`